### PR TITLE
Refactor parser, add tests, fix issues

### PR DESCRIPTION
**Pull request by kirillmamaev**
## 📝 What does this PR do?
- Refactor parser
- Add tests
- Fix issues with missing spaces
- Update JSON file

### CODE REVIEW

File: pdf_parsing/parser.py [hunk 1]
```diff
@@ -2,102 +2,152 @@
...
-    return date, title, quote, quote_source, explanation
-
-
-def extract_text(pdf_file, start, end, breaks):
-    entries = {}
-
-    with fitz.open(pdf_file) as f:
-        for page in f:
-            if page.number < start:
+PARSING_FOLDER = "./pdf_parsing/"
```
Comment (bbelderbos) [inline pdf_parsing/parser.py]: I recommend using pathlib.Path here

File: pdf_parsing/parser.py [hunk 2]
```diff
@@ -2,102 +2,152 @@
...
+END_QUOTE_SYMBOL = "\u201d"  # Origin: ”
+START_SOURCE_SYMBOL = "\u2014"  # Origin: —
+
+
+# Parses The Daily Stoic book.
+# The book is a PDF file with a specific structure.
+class Parser:
+    # Returns a list of all pdf files in a folder.
+    def get_pdfs(self, folder="."):
+        return [folder + "/" + file for file in os.listdir(folder) if ".pdf" in file]
```
Comment (bbelderbos) [inline pdf_parsing/parser.py]: and here

File: pdf_parsing/parser.py [hunk 3]
```diff
@@ -2,102 +2,152 @@
...
+                    "quote": quote,
+                    "quote_source": quote_source,
+                    "explanation": explanation,
+                }
+
+        return entries
+
+    # Extracts the date, source, quote and explanation from a page of the Daily Stoic
+    def __parse_page_text(self, text: str):
+        lines = text.replace("\t", " ").split("\n")
```
Comment (bbelderbos) [inline pdf_parsing/parser.py]: you can also use `splitlines`

File: pdf_parsing/parser.py [hunk 4]
```diff
@@ -2,102 +2,152 @@
...
+                    "title": title,
+                    "quote": quote,
+                    "quote_source": quote_source,
+                    "explanation": explanation,
+                }
+
+        return entries
+
+    # Extracts the date, source, quote and explanation from a page of the Daily Stoic
+    def __parse_page_text(self, text: str):
```
Comment (bbelderbos) [inline pdf_parsing/parser.py]: I am used to single underscore for 'private' functions/methods, any reason to use __?

File: pdf_parsing/parser.py [hunk 5]
```diff
@@ -2,102 +2,152 @@
...
-    data = extract_text("The Daily Stoic.pdf", start=start_page, end=end_page, breaks=skip)
+        return explanation_first_letter, date, title, lines[index:]
+
+    # Checks if the current line is the end of quote.
+    def __is_end_of_quote(self, line, next_line):
+        # This is required to handle cases where the END_QUOTE_SYMBOL is missing
+        return (
+            line[-1] == END_QUOTE_SYMBOL
+            or (next_line[0] == START_SOURCE_SYMBOL and next_line[1:6].isupper())
+            or next_line[0:5].isupper()
```
Comment (bbelderbos) [inline pdf_parsing/parser.py]: seems this magic number is repeated, so you could also wrap that into a constant

File: pdf_parsing/parser.py [hunk 6]
```diff
@@ -2,102 +2,152 @@
...
+    298,
+    330,
+    361,
+]
+START_QUOTE_SYMBOL = "\u201c"  # Origin: “
+END_QUOTE_SYMBOL = "\u201d"  # Origin: ”
+START_SOURCE_SYMBOL = "\u2014"  # Origin: —
+
+
+# Parses The Daily Stoic book.
```
Comment (bbelderbos) [inline pdf_parsing/parser.py]: you could turn this into a class docstring

File: pdf_parsing/parser.py [hunk 7]
```diff
@@ -2,102 +2,152 @@
...
+# The book is a PDF file with a specific structure.
+class Parser:
+    # Returns a list of all pdf files in a folder.
```
Comment (bbelderbos) [inline pdf_parsing/parser.py]: this comment seems redundant and/or you could make the method name below more descriptive

File: pdf_parsing/parser.py [hunk 8]
```diff
@@ -2,102 +2,152 @@
...
+    def get_pdfs(self, folder="."):
+        return [folder + "/" + file for file in os.listdir(folder) if ".pdf" in file]
+
+    # Parses The Daily Stoic book and returns a dictionary with the entries.
```
Comment (bbelderbos) [inline pdf_parsing/parser.py]: docstring

File: pdf_parsing/parser.py [hunk 9]
```diff
@@ -2,102 +2,152 @@
...
-                "explanation": explanation
-            }
+            explanation_lines.append(line)
 
-    return entries
+        date = date[:-2]
+        quote = self.__apply_quotes(" ".join(quote_lines).strip())
+        explanation = (explanation_first_letter + " ".join(explanation_lines)).strip()
 
+        return date, title, quote, quote_source, explanation
```
Comment (bbelderbos) [inline pdf_parsing/parser.py]: could use a named tuple here (and add a return type hint)

File: pdf_parsing/parser.py [hunk 10]
```diff
@@ -2,102 +2,152 @@
...
 
-if __name__ == "__main__":
+    # Extracts the first letter of explanation, date and title from the first lines of the page
```
Comment (bbelderbos) [inline pdf_parsing/parser.py]: docstring

File: pdf_parsing/parser.py [hunk 11]
```diff
@@ -2,102 +2,152 @@
...
+        date = lines[index].strip()
+        index += 1
 
-    # These are specific to Ishan's copy of the Daily Stoic
-    start_page = 14
-    end_page = 392
-    skip = [45, 75, 107, 138, 139, 171, 202, 234, 266, 267, 298, 330, 361]
+        # Required to cover cases where the title is split into multiple lines.
+        title_lines = []
+        while lines[index].isupper():
```
Comment (bbelderbos) [inline pdf_parsing/parser.py:208]: what is this condition / check for? I wonder if a list comprehension could be simpler here

File: pdf_parsing/parser.py [hunk 12]
```diff
@@ -2,102 +2,152 @@
...
+            title_lines.append(lines[index])
+            index += 1
+        title = "".join(title_lines).strip()
 
-    data = extract_text("The Daily Stoic.pdf", start=start_page, end=end_page, breaks=skip)
+        return explanation_first_letter, date, title, lines[index:]
```
Comment (bbelderbos) [inline pdf_parsing/parser.py]: same here, a named tuple as a way of grouping variables could make the return simpler here

File: pdf_parsing/test_parser_result.py [hunk 13]
```diff
@@ -0,0 +1,300 @@
+import json
+import unittest
```
Comment (bbelderbos) [inline pdf_parsing/test_parser_result.py]: Thanks for adding tests, have you considered using pytest?

**Review by bbelderbos — APPROVED** [review]
Thanks for this PR / your work, appreciated! 

I added some feedback including some stylistic things / preferences so feel free to add/adjust what you see necessary. Overall this looks good to me. 

I'll let you @IshanMi review it as well and merge.

File: pdf_parsing/parser.py (hunk 4 above)
Comment (kirillmamaev) [reply pdf_parsing/parser.py]: No reason to use `__` over `_`, beside my own rusty Python skills. 😄
Replaced.

File: pdf_parsing/test_parser_result.py (hunk 13 above)
Comment (kirillmamaev) [reply pdf_parsing/test_parser_result.py]: Rewrote tests to use `pytest` instead of `unittest`.

**Comment by kirillmamaev** [comment]
Hi @IshanMi 👋  Please review and merge if happy!
@bbelderbos Thanks for your review and comments! 🙌

File: pdf_parsing/test_parser_result.py (hunk 13 above)
Comment (bbelderbos) [reply pdf_parsing/test_parser_result.py]: wow awesome, you did not have to ;)
but I hope you felt the joy of using pytest :)

**Comment by bbelderbos** [comment]
Wow thanks, I am happy to merge this tomorrow ... 
We can always do a follow up. 

@mamaevdev curious, are you using this somewhere?

**Comment by kirillmamaev** [comment]
> @mamaevdev curious, are you using this somewhere?

We use it to post Daily Stoic into our team channel in Slack. Missing spaces annoyed me a bit... 😄

**Comment by bbelderbos** [comment]
That's awesome! 

We moved our community to Circle, maybe nice to figure out how to port this to that platform as well :)

**Pull request merged**